lib/poolex.ex

```diff
@@ -157,28 +151,28 @@ defmodule Poolex do
   Same as `run!/3` but handles runtime_errors.
 
   Returns:
-    * `{:runtime_error, reason}` on errors.
-    * `:all_workers_are_busy` if no free worker was found before the timeout.
+    * `{:error, {:runtime_error, reason}}` on errors.
+    * `{:error, :checkout_timeout}` if no free worker was found before the timeout.
 
   See `run!/3` for more information.
 
   ## Examples
 
       iex> Poolex.start_link(pool_id: :some_pool, worker_module: Agent, worker_args: [fn -> 5 end], workers_count: 1)
       iex> Poolex.run(:some_pool, fn _pid -> raise RuntimeError end)
-      {:runtime_error, %RuntimeError{message: "runtime error"}}
+      {:error, {:runtime_error, %RuntimeError{message: "runtime error"}}}
       iex> Poolex.run(:some_pool, fn pid -> Agent.get(pid, &(&1)) end)
       {:ok, 5}
   """
   @spec run(pool_id(), (worker :: pid() -> any()), list(run_option())) ::
-          {:ok, any()} | :all_workers_are_busy | {:runtime_error, any()}
+          {:ok, any()} | {:error, :checkout_timeout | {:runtime_error, any()}}
   def run(pool_id, fun, options \\ []) do
     {:ok, run!(pool_id, fun, options)}
   rescue
-    runtime_error -> {:runtime_error, runtime_error}
+    CheckoutTimeoutError -> {:error, :checkout_timeout}
+    runtime_error -> {:error, {:runtime_error, runtime_error}}
```

Comment:
As I've already told you in this comment, I don't see the point of having two `run` & `run!` functions https://github.com/general-CbIC/poolex/issues/51#issuecomment-1637804086

Here, we are using `raise` to handle the control flow. It is not a good practice in Elixir. First, we capture a timeout to raise a `CheckoutTimeoutError`, and later we capture that error to return a tuple. If you still want to maintain both function I would do the opposite, I implement `run`, and the `run!` should use `run` to implement it self. This way you avoid this double raise.

Reply:
It looks like it's time to merge the interfaces :)

I will do it in #61